### PR TITLE
[action] fix contract addr in tsf

### PR DESCRIPTION
**Pull request by Liuhaai**
The contractAddr should be empty if the recipient is not contract

### CODE REVIEW

File: action/protocol/account/transfer.go [hunk 1]
```diff
@@ -127,11 +127,10 @@ func (p *Protocol) handleTransfer(ctx context.Context, act action.Action, sm pro
 	}
 
 	receipt := &action.Receipt{
-		Status:          uint64(iotextypes.ReceiptStatus_Success),
-		BlockHeight:     blkCtx.BlockHeight,
-		ActionHash:      actionCtx.ActionHash,
-		GasConsumed:     actionCtx.IntrinsicGas,
-		ContractAddress: p.addr.String(),
```
Comment (millken) [inline action/protocol/account/transfer.go:134]: Why we removed it?

Comment (dustinxie) [reply action/protocol/account/transfer.go:134]: need this, otherwise the receipt root will change

File: action/protocol/rewarding/protocol.go [hunk 2]
```diff
@@ -68,6 +68,16 @@ func NewProtocol(cfg genesis.Rewarding) *Protocol {
 	}
 }
 
+// ProtocolAddr returns the address generated from protocol id
+func ProtocolAddr() address.Address {
```
Comment (dustinxie) [inline action/protocol/rewarding/protocol.go:72]: create a common func `hashStringToAddress()` under `/action/protocol`
and here
```
return hashStringToAddress(protocolID)
```

File: action/protocol/rolldpos/epoch.go [hunk 3]
```diff
@@ -92,6 +95,16 @@ func NewProtocol(numCandidateDelegates, numDelegates, numSubEpochs uint64, opts
 	return p
 }
 
+// ProtocolAddr returns the address generated from protocol id
+func ProtocolAddr() address.Address {
```
Comment (dustinxie) [inline action/protocol/rolldpos/epoch.go:99]: same here

File: api/web3server_utils.go [hunk 4]
```diff
@@ -239,7 +244,7 @@ func (svr *Web3Server) getTransactionFromActionInfo(actInfo *iotexapi.ActionInfo
 			if err != nil {
 				return transactionObject{}, err
 			}
-			addr, err := ioAddrToEthAddr(receipt.ContractAddress)
+			addr, err := getExecutionContractAddr(receipt.ContractAddress)
```
Comment (dustinxie) [inline api/web3server_utils.go:247]: in `ioAddrToEthAddr`, for `input = ""`
```
return "0x0000000000000000000000000000000000000000", nil
``` 
now in `getExecutionContractAddr`, for `input = ""`
```
return "", nil
```
would this be an issue (behavior change for client)?

File: api/web3server_utils.go [hunk 5]
```diff
@@ -372,6 +377,22 @@ func (svr *Web3Server) getLogsWithFilter(from uint64, to uint64, addrs []string,
 	return ret, nil
 }
 
+// contract addr is only returned when a contract was created, otherwise an empty string is returned
+func getExecutionContractAddr(addr string) (string, error) {
+	switch addr {
+	case "":
+		fallthrough
+	case account.ProtocolAddr().String(),
```
Comment (dustinxie) [inline api/web3server_utils.go:385]: you can just combine `""` into here

File: action/protocol/rewarding/protocol.go (hunk 2 above)
Comment (Liuhaai) [reply action/protocol/rewarding/protocol.go:72]: I don't think exporting protocolID is better

File: api/web3server_utils.go (hunk 4 above)
Comment (Liuhaai) [reply api/web3server_utils.go:247]: Good finding. But I think there will never be a scenario where both  `to` and `contractAddr` are empty string

**Review by CoderZhi — APPROVED** [review]
the others look good to me

File: api/web3server_utils.go [hunk 6]
```diff
@@ -374,6 +379,21 @@ func (svr *Web3Server) getLogsWithFilter(from uint64, to uint64, addrs []string,
...
 
+// contract addr is only returned when a contract was created, otherwise an empty string is returned
+func getExecutionContractAddr(addr string) (string, error) {
+	switch addr {
+	case "",
+		account.ProtocolAddr().String(),
+		poll.ProtocolAddr().String(),
+		rewarding.ProtocolAddr().String(),
+		rolldpos.ProtocolAddr().String(),
+		staking.ProtocolAddr().String():
```
[the remark this answers is not in the document]
Comment (Liuhaai) [reply api/web3server_utils.go:390]: The addr `protocol.HashStringToAddress(protocolID)` is used in returned Receipt object

**Review by dustinxie — APPROVED** [review]

**Pull request merged**